subsys/usb/device/usb_transfer.c

```diff
@@ -333,6 +333,10 @@ int usb_transfer_sync(uint8_t ep, uint8_t *data, size_t dlen, unsigned int flags
 
 		trans = usb_ep_get_transfer(ep);
 		if (!trans || trans->status != -EBUSY) {
+			ret = k_sem_take(&pdata.sem, K_MSEC(USB_TRANSFER_SYNC_TIMEOUT));
```

Comment:
What is the difference between this and just doubling `USB_TRANSFER_SYNC_TIMEOUT`?

Reply:
>I have observed this happen in a system that was heavily slowed down
by hundreds of IMMEDIATE uart log messages.

Apart from incorrect logic to take it again, other semaphores and everything else that has timeouts in the whole system might expire because of "IMMEDIATE uart log messages".

Reply:
@tmon-nordic 

> What is the difference between this and just doubling `USB_TRANSFER_SYNC_TIMEOUT`?

I have not tried this, but I suspect doubling USB_TRANSFER_SYNC_TIMEOUT from 100ms to 200ms will make the probability of hitting this error 50% of what it was before.

Note that the timeout in use is not an actual timeout for the usb_transfer_sync() call, but an intermediate timeout that effectively sets the polling rate to detect the condition when the transfer has been canceled. Notice that k_sem_take is called in a loop.

So completion of the transfer will be detected quickly when semaphore is released, but a cancellation of the transfer can only be detected every USB_TRANSFER_SYNC_TIMEOUT period.

@jfischer-no 

> Apart from incorrect logic to take it again, other semaphores and everything else that has timeouts in the whole system might expire because of "IMMEDIATE uart log messages".

Could you clarify which logic you think is wrong and why? I think it's perfectly fine to try to take the semaphore again, as long as the previous attempt to take it did not succeed. This can happen if the transfer takes a long time to complete.  Notice that if the semaphore is taken successfully we break out of the loop, and we don't attempt to take it again. I don't see anything wrong with this logic.

The problem I am fixing here is a race condition, whereby the transfer is still pending, when the k_sem_take on line 329 returns, then context switching or an interrupt happens, before the transfer status is checked, and then the transfer suddenly completes successfully, and its status will be updated to not be EBUSY. When we return to our thread, the status check will fail, and our code will think that the transfer has failed, when in fact it has completed successfully. That's why trying to take it just one more time fixes the issue, cause  if taking it succeeds, it means it has completed, and if taking it fails, then we know the above check caught a real cancellation.

Regarding the IMMEDIATE uart messages, absolutely, things will start expiring, and it's a great way to accidentally test for subtle race condition bugs such as this one. Another good way to test would be to write a timer interrupt handler that periodically locks up the CPU in ISR context for long times, and see where things break. In some cases it will show things that are not real errors and on a real system the condition would never be triggered, and in other cases it might show real bugs that have a very low probability of occurring.

In this case I believe this is a real bug, because USB transfers can sometimes take a long time to complete. If the system is not loaded down with long-running ISRs, then it will be an extremely rare event that the completion of the USB transfer lines up with the expiration of the k_sem_take() call, but it can still happen, once in a blue moon, therefore it's a real bug.

Reply:
@jfischer-no 

I think a fundamental reason for our disagreement comes from us not agreeing about the purpose of USB_TRANSFER_SYNC_TIMEOUT. I believe its purpose is not to set a timeout for the usb_transfer_sync() call, but to set the periodicity at which this function checks for cancellation. I don't believe usb_transfer_sync() should _ever_ time out on its own. Maybe there could be a different timeout in the users of this code, which when expires causes the transfer to be canceled, but that would be an independent, unrelated timeout.

To support my position, I have looked up when this timeout mechanism has been implemented, take a look at commit cd7461495f1aeb4561d6a7e96c9f37b029805694 , and it's description. "The fix introduces a periodic check if a transfer is still valid."

@xhpohanka Please comment here if you agree or not.

> Transfer can be cancelled by usb_cancel_transfer() for whatever reason, would you change not take semaphore for nothing?

I don't think it's for no reason.
First of all taking it again will not fundamentally wrong, because the semaphore lives on the stack and it has not gone out of scope.
It is possible that no other thread or ISR is planning anymore to be giving that semaphore, and this does indeed happen if the transfer is canceled, but then the timeout functionality of the semaphore will still work, so what it can cause is an added delay equal to the timeout when it comes to detecting canceled transfers.
Now it is possible that the transfer completes right after the previous k_sem_take() times out, and in that case this k_sem_give() will return without having to wait for a timeout, and that is a sign that the transfer has actually completed, even though the EBUSY check failed, so the transfer has not been canceled. This is exactly what I have observed in my testing.
So on line 336 we're in a state in which k_sem_take() could either time out or could actually take the semaphore, which is not any different from the very first call to k_sem_take(), therefore it's not wrong.

> It does not matter for usb_transfer_sync(), it was not completed within USB_TRANSFER_SYNC_TIMEOUT. I will also not go into any usb_transfer design flaws here.

I didn't understand what you meant by this sentence.

> If you think USB_TRANSFER_SYNC_TIMEOUT is too small, then please move it to Kconfig option.

If anything I think USB_TRANSFER_SYNC_TIMEOUT might be too big, not too small. I would support reducing it, because it would increase the frequency at which the code would check for canceled transfers, and we'd be able to detect a cancellation sooner.

On a different note, I think I see something else worrying:
in `usb_transfer_work()` the transfer semaphore is given on `done:`, before calling the callback.
That means that another thread wishing to transfer on the same endpoint, could potentially have started another transfer, cause `usb_transfer_is_busy(uint8_t ep)` becomes false. So then when checking for cancellation in `usb_transfer_sync`, we could find the _new_ transfer and think that it is the one that the running `usb_transfer_sync` started. So it could be going back looping and waiting on its own semaphore that is never gonna be given, and only return when no new transfers will be pending anymore on the same endpoint.

So I actually think that there is a design flaw here that makes it impossible for usb_transfer_sync to detect cancellation without false negatives. My changeset in this PR only fixes false positives. A more complete solution that fixes both false positives and false negatives I think would require a slight redesign: Either it should be possible to setup a callback that is called on cancellation, or perhaps it should be possible to set up a transfer that is not released automatically once done, and will prevent new transfers to the same endpoint until usb_transfer_sync can manually cancel it. Not sure what the best way to do it would be, but I think it can be done without making backwards incompatible changes to the usb_transfer api.
For the purpose of discussion I will try to implement an alternate solution that will do both, but may take me some time to get it tested.

Reply:
Alright, so I have implemented the alternate, better solution https://github.com/zephyrproject-rtos/zephyr/pull/54743 and I have marked it as a draft for now for the purpose of discussion.

Note that I have not done any directed testing on it. I have tested that it compiles, and it doesn't break my keyboard firmware, but I have not tried triggering the race conditions this tries to fix, so I have marked it as a draft.

Reply:
> To support my position, I have looked up when this timeout mechanism has been implemented, take a look at commit https://github.com/zephyrproject-rtos/zephyr/commit/cd7461495f1aeb4561d6a7e96c9f37b029805694 , and it's description. "The fix introduces a periodic check if a transfer is still valid."

>  Please comment here if you agree or not.

Hello. I have implemented the original solution because I had problems with occasional deadlocks on that semaphore in my STM32 project. I'm not an expert on zephyr USB stack and the periodic check was the first simple solution. In my pull request I have stated that the `USB_TRANSFER_SYNC_TIMEOUT` value was not tested and I agree that it might not be optimal.

Your solution with another callback might be even better, but maybe it is becoming too complicated. Wouldn't it work with additional callback parameter?

Reply:
usb_transfer() appears to be a public API. Adding an additional callback parameter would be a backwards-incompatible change, so I think the second callback is necessary.

Reply:
@jfischer-no  Can you comment on what you think of https://github.com/zephyrproject-rtos/zephyr/pull/54743 ? Have I succeeded in convincing you that there is a real race condition here?
If you think the other PR is going in the right direction, I can close this one, and clean it up #54743 to pass compliance checks, and remove the draft label.

Reply:
>Have I succeeded in convincing you that there is a real race condition here?

Well, not really. It behaves how it is designed. 

> > What is the difference between this and just doubling `USB_TRANSFER_SYNC_TIMEOUT`?
> 
> I have not tried this, but I suspect doubling USB_TRANSFER_SYNC_TIMEOUT from 100ms to 200ms will make the probability of hitting this error 50% of what it was before.

I do not think it scales the way you assume. Please try to increase USB_TRANSFER_SYNC_TIMEOUT.